oeps/oep-0049-django-app-patterns.rst

```diff
@@ -0,0 +1,221 @@
+
+.. _OEP-0019: https://open-edx-proposals.readthedocs.io/en/latest/oep-0019-bp-developer-documentation.html#readmes
+
+.. ___init__.py:
+__init__.py
++++++++++++
+The ``__init__.py`` file should contain a line for the ``default_app_config`` for the app, or o. This ``default_app_config`` should point to the ``AppConfig`` located in ``<app_name>/apps.py``. It may also contain small app details such as a version. However, unlike many packages, ``__init__.py`` should *not* be used to as the way to export the app's public methods. These should be exported using ``api.py`` (and thus imported as ``from path.to.app.api import public_function``). See api.py_ below.
+
+For example:
+
+.. code-block:: python
+
+  default_app_config = "service_name.apps.app_name.apps.CustomAppConfig"
+
+
+.. _apps.py:
+apps.py
++++++++
+The ``apps.py`` file should contain a subclass of a Django ``AppConfig``. The AppConfig should set the app's name to its full path (e.g. ``name = "service_name.apps.app_name"``) and should (optionally) have an overriding ``ready()`` function which initializes the app. Any imports that need to happen during app initialization (such as signals_) need to happen inside the ``ready`` function or else there's risk of circular imports.
+
+For example:
+
+.. code-block:: python
+
+  class MyAppConfig(AppConfig):
+      """
+      Application Configuration for MyApp.
+      """
+      name = "service_name.apps.app_name"
+
+      # (optional) Set up plugin. See https://github.com/edx/edx-django-utils/tree/master/edx_django_utils/plugins
+
+      def ready(self):
+          """
+          Connect handlers to recalculate grades.
+          """
+          from .signals import handlers
+
+.. _api.py:
+api.py
+++++++
+This should be single point of entry for other Python code to talk to your app. This is *not* a Rest API, this is a Python API (see rest_api_). Some rules for ``api.py`` are as follows:
+
+1. API methods defined in ``api.py`` should be well-named, self-consistent, and relevant to its own domain (without exposing technical and implementation details)
+2. An app's Django models and other internal data structures should not be exposed via its Python APIs.
+
+
+Not exposing an app's data structures can be tricky because it's very easy to expose them without meaning to. Therefore there are a couple common strategies we employ.
+
+1. When importing internal app code to be used in the ``api.py`` file, prefix it with an underscore so it's clear it's for internal use only.
+
+2. Create a ``data.py`` file to house simple data objects that can be passed from your app's function to the calling app. By creating these objects, we can avoid both passing Django model objects or querysets directly and having to serialize data. Other apps may import data classes from ``data.py`` in additional to functionality from ``api.py``. See data.py_ for more details.
```

Comment:
I'm concerned about not exposing QuerySets, due to future performance concerns, but also due to extensibility. For instance, I was just discussing w/ @MichaelRoytman about the course_enrollment api in edx-platform which doesn't currently expose a QuerySet, and as a result he is unable to additionally filter those enrollments inside the database (and instead had to pull the data into memory and then filter in python). That pattern, of pulling into memory and filtering, makes pagination difficult, because to find a page of N users, you may have to filter through the entire data set in python (in the worst case). Even exposing generators just mitigates the problem by only pulling users until you fill up the filtered page, but that doesn't prevent you from needing to iterate the users in python, rather than filtering in the database.

Reply:
(Context: this was mainly pulled from https://github.com/edx/edx-platform/blob/master/docs/decisions/0002-inter-app-apis.rst#decisions.) 

I've also been hesitant about performance on this point. There are ways around the performance issue, such as allowing `kwargs` that get passed to the initial queryset, but that makes the hard line we're trying to draw between apps a lot more hazy. It also defeats the purpose of not exposing the internal data structure, but it'd give us a performant API that also only returns plain data classes. 

There was a recent discussion with about `models_api.py` (which I had pulled from this doc), which set it up as the "when you need to break the API contract for performance reasons" API to use. I'm not sure if having a separate file, or having kwargs (above) would be less hazy. 

cc: @nasthagiri since you wrote the referenced ADR and had that `models_api.py` discussion with me.

Reply:
Even exposing **kwargs doesn't actually help that much. For instance, you couldn't use `.annotate` or `.aggregate` if the result isn't returning a QuerySet. I wonder if there's a path forward where we advocate returning `values()` or `values_list()` querysets. Even that, though, means that clients can't choose which columns they care about, which isn't great from performance purposes.

Sigh... I wish we had a better way of being able to use the Python standard of "we're all consenting adults" to feel more comfortable exposing honest-to-goodness querysets. Like, can we just say that non-public fields on a model will be prefixed with `_` on the object (and then we can name the column whatever we want on the field definition)

Reply:
So coincidentally, I had something like this come up in the ticket that I'm working on today, and my mind also wandered to trying to address it with `values_list`. The problem was:

* I wanted to return an iterable of course contexts that didn't have any learning_sequence outline data associated with them.
* Because I want to punt `learning_sequences` out of edx-platform and into its own repo, I didn't want to make it know about CourseOverview.
* The actual pushing of data into `learning_sequences` happens in contentstore (i.e. Studio), since `learning_sequences` likewise doesn't know about modulestore.

So I settled on a new addition to the `learning_sequences` API that was:

```python
def get_course_keys_with_outlines() -> QuerySet:
    """Queryset of ContextKeys, iterable as a flat list."""
    return LearningContext.objects.values_list('context_key', flat=True)
```

So iterating through it simply goes like you'd expect:

```python
for course_key in get_course_keys_with_outlines():
    print(course_key)
```

Yields:

```
course-v1:TNL+100+2021_01_21
course-v1:TNL+100+ABC
course-v1:TNL+LibContentTest+2021
course-v1:edX+DemoX+Demo_Course
```

But you can also use it as a qset:

```python
all_course_keys_qs = CourseOverview.objects.values_list('id', flat=True)
missing_outlines_qs = all_course_keys_qs.exclude(
    id__in=get_course_keys_with_outlines()
)
```

The query it uses is:

```
SELECT `course_overviews_courseoverview`.`id` FROM `course_overviews_courseoverview`
WHERE NOT (`course_overviews_courseoverview`.`id` IN (SELECT U0.`context_key` FROM `learning_sequences_learningcontext` U0))
```

Not that this will solve every problem, but I think there is something we can do here that is to encourage some predictable subset of the Queryset API in some disciplined way.

Reply:
I mean, would it work if we subclassed QuerySet to make something that maps into attrs over it when you actually do the iteration? I'll experiment a little...

Reply:
Thinking on this a little more, let's lay out some use cases for needing to send back a queryset in an API. Some things off the top of my head:

1. Queryset of simple objects like the ContextKeys in the example above, where it's an input going into a larger query.
2. Queryset of objects we want for something like paginated responses, including filtering–used in REST API views.
3. Queryset where a client has some data associated with an entity in another app. So if I'm some new grading system, I might have a 1:M relationship between a `LearningSequence` and my `UserSequenceGrade` table or some such.
4. Similar to (3), but a variant where you don't care about any of the actual extra data on the other end of your relationship, but you just need to be able to associate that data by IDs (e.g. tacking on purchase/upgrade related information to enrollments).

I do think it's fine to not have to optimize all the way down. So for instance, we should definitely avoid the whole N+1, "A new query for every row" problem, but at the same time, having one call per system on that view should be fine. So if I have an API endpoint or page that returns Enrollments, I think it's okay if the query for for `select id from student_courseenrollment...` gets run once for whatever is adding the purchase data, a second time for whatever is displaying course metadata, etc. We could make it a pattern that goes like this:

* App A has a model that has a foreign key constraint on a model in app B.
* App A's API provides a function that takes as an argument a QuerySet of app B entities (really, it just needs to be their IDs), and then returns a QuerySet that has an `attr` data class from App A's `data.py` for things attached to that. I think we could do this by using a custom manager and a subclass of QuerySet, but I don't have time to really check that tonight.

I realize that only covers M:1 and 1:1 relationships and I haven't really got a good story for 1:M and M:M...? Going to mull it over more tomorrow.

Reply:
Based on these comments it seems less likely that we'll be able to have a hard line on `api.py` only returning data classes (which is fine). It also sounds like we might be able to iterate on a future solution where we get both abstractions and performance, but I think we'd like to see those solutions battle tested rather than doc-first. 

In the interim time, where do we think the "good enough for most people, break the rules if you have to" compromise lives? Options I see thus far:
1. Have `api.py` return data classes and have `models_api.py` return querysets
2. Leave everything in `api.py` and more generly suggest data classes over querysets
3. Leave everything in `api.py` and not suggest a preference on data return types 
4. Some better idea you have than 1-3.

Reply:
Yeah, I don't think we can hold a hard line at the moment. I've been playing around with this a little over the weekend, and I think that we _might_ be able to have our cake and eat it as well, with something like this:

```python

# Assuming there's a TaskData class corresponding to a Task model..

def data_from_model(data_cls, model):
    model_data = {a.name: getattr(model, a.name) for a in attr.fields(data_cls)}
    return data_cls(**model_data)

class TaskDataQuerySet(models.QuerySet):
    def __iter__(self):
        for model in super().__iter__():
            yield data_from_model(TaskData, model)

class Task(models.Model):
    objects = TaskDataQuerySet.as_manager()
    # All the other fields...
```

Then you get a queryset that you can still do query manipulations on (and we can customize the manager to make for more natural queries), but when you iterate it, you get the attrs data structures instead of mutable models.

Parameterizing this stuff was trickier than I thought because QuerySets and Managers expect to be initialized in a kind of magical way deep inside the ORM for certain operations, so I think we'd have to rely on subclassing rather than passing in the class as a constructor param.

I want to try to sit down and really think about the implications of this approach, I don't think I'm going to have time for at least another week or two. :-(

Reply:
That looks interesting and similar to how I thought we'd do performant paginators. I'll extend the deadline so we can hash this out. I'd rather it be right than done.

Reply:
So I think if we want that balance of performance but predictability, it might look something like:

* There's a `DataQuerySet` and `DataQueryManager` in a library we have somewhere.
* People subclass them when they're putting together new models and put class attributes like the data class, conversion specifics, list of attributes you're allowed to sort by, custom manager methods, etc.
* We attach that as something other than the default manager to the model, so that we don't break Django admin and such. Maybe call it `read_only`, if that's what it's effectively going to be?
* Managers and Querysets give us a place to make some flexible querying logic that is an approved subset of the general ORM facilities. We could even make it part of the public API, with the understanding that the search patterns explicitly on those classes are approved, and if you do anything else, you're playing with fire. We're all adults and all, so we don't need to lock things down, just strongly signal.

Reply:
So for the last part, something might look like:

```python
tasks = api.get_tasks_for_user(user_id)

urgent_tasks = tasks.overdue().high_priority()
```

Reply:
(Sorry, still doing this in bits and pieces where I can fit in the time...)

I think the kind of interface I posted above could work for the "there is a system out there that I want answers from" type of query, but we'd need something more for the "I want to efficiently add my own data to this" type of query.

Say I was attaching course-specific survey data to enrollments (I am picking totally at random). There's an enrollments API built in the fashion of this OEP, and I have survey objects that have a reverse relationship name of "survey". The approach I suggested above would obliterate the idea of a model relation, and I'm super-skittish about reimplementing that type of logic. But a client _should_ be able to do something like:

```python

tasks = task_api.get_tasks_for_user(user_id)
urgent_tasks = tasks.overdue().high_priority()

jira_tickets = jira_api.tickets_for_tasks(urgent_tasks)
```

Implementing the JIRA API `tickets_for_tasks` by taking a queryset, doing a `select_related`, and returning a matching qset with JIRA information corresponding to the qset that was passed in–that general pattern _should_ be a common use case, right? The JIRA app here gets to determine the reverse relationship name in their model...

Reply:
(FYI I extended the review date by a month so we could hash this out. New date is April 10.)

I really like the "have your cake and eat it too" design. I have a lot of hesitations on adding any of it in here though. Reasons being:
1. Whatever we decide on will be an idea and not tested in multiple places. I think this should be an `ideate -> implement -> ADR -> OEP` flow, rather than tagged on here. (Very interested in this topic though, so I'd be happy to be a part of the future work)
2. Sort of in regard to "reimplementing that type of logic", this feels like it's straddling the line between a customization and a contradiction to how Django does things. I don't think we should bend to Django's will, but as one of the larger Django projects, I think this may be a good space for cross-community collaboration to see if this is something someone's already tried, or if they have any thoughts. Honestly it feels like it'd be a rabbit hole to dive into. 

TL;DR: I want this; I want to be in on the discussion; I don't think this is where we decide it.

I'm thinking this section might become more of a "use attrs if possible, querysets are okay for performance reasons".  When we have a working use case, we can update this section to use it. 

That seem fair to you? I'll update the PR with that language if so. 

P.S. I hate stopping a brainstorm, so know that I feel bad about it. Haha.

Reply:
Sorry I didn't reply to this. I agree that this idea is too rough to be part of this OEP, and it makes sense to experiment with it first.

Reply:
Thanks! I've updated the language to be less strict about not passing querysets. If you have any thoughts on it, let me know. If not, I think this was the last major question, so we'll just wait it out until April 10 :)